File: src-tauri/src/gemini.rs
use anyhow::Result;
use reqwest::Client;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
struct GeminiRequest {
    contents: Vec<Content>,
}

#[derive(Debug, Serialize, Deserialize)]
struct Content {
    parts: Vec<Part>,
}

#[derive(Debug, Serialize, Deserialize)]
struct Part {
    text: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct GeminiResponse {
    candidates: Vec<Candidate>,
}

#[derive(Debug, Serialize, Deserialize)]
struct Candidate {
    content: Content,
}

pub async fn send_message_to_gemini(
    api_key: &str,
    model: &str,
    message: &str,
) -> Result<String> {
    let client = Client::new();
    
    // Map model name to full Gemini model ID
    // Using the stable Gemini models that are available
    let model_id = match model {
        "2.5-flash" => "gemini-1.5-flash",  // Flash variant
        "2.5-pro" => "gemini-1.5-pro",      // Pro variant
        "3-flash" => "gemini-1.5-flash",    // Using 1.5-flash for now
        "3-pro" => "gemini-1.5-pro",        // Using 1.5-pro for now
        _ => "gemini-1.5-flash",            // Default to flash
    };
    
    let url = format!(
        "https://generativelanguage.googleapis.com/v1/models/{}:generateContent?key={}",
        model_id, api_key
    );
    
    let request_body = GeminiRequest {
        contents: vec![Content {
            parts: vec![Part {
                text: message.to_string(),
            }],
        }],
    };
    
    let response = client
        .post(&url)
        .json(&request_body)
        .send()
        .await?;
    
    if !response.status().is_success() {
        let error_text = response.text().await?;
        return Err(anyhow::anyhow!("Gemini API error: {}", error_text));
    }
    
    let gemini_response: GeminiResponse = response.json().await?;
    
    if let Some(candidate) = gemini_response.candidates.first() {
        if let Some(part) = candidate.content.parts.first() {
            return Ok(part.text.clone());
        }
    }
    
    Err(anyhow::anyhow!("No response from Gemini API"))
}

File: python-lib/kiassist_utils/gemini.py
"""Gemini API integration module."""

from google import genai
from google.genai import errors
from typing import Optional


class GeminiAPI:
    """Handle interactions with Google Gemini API."""
    
    # Model mapping
    MODEL_MAP = {
        "2.5-flash": "gemini-1.5-flash",
        "2.5-pro": "gemini-1.5-pro",
        "3-flash": "gemini-1.5-flash",  # Using 1.5-flash for now
        "3-pro": "gemini-1.5-pro",      # Using 1.5-pro for now
    }
    
    def __init__(self, api_key: str):
        """Initialize Gemini API client.
        
        Args:
            api_key: Google Gemini API key
        """
        self.client = genai.Client(api_key=api_key)
    
    def send_message(self, message: str, model: str = "2.5-flash") -> str:
        """Send a message to Gemini and get a response.
        
        Args:
            message: The message to send
            model: The model identifier (2.5-flash, 2.5-pro, etc.)
            
        Returns:
            The response text from Gemini
            
        Raises:
            Exception: If the API call fails
        """
        # Map model name to full Gemini model ID
        model_id = self.MODEL_MAP.get(model, "gemini-1.5-flash")
        
        try:
            response = self.client.models.generate_content(
                model=model_id,
                contents=message
            )
            return response.text
            
        except errors.APIError as e:
            raise Exception(f"Gemini API error: {str(e)}")
        except Exception as e:
            raise Exception(f"Unexpected error: {str(e)}")
